### PR TITLE
Replace CountDownLatches for response handlers with CompletableFutures

### PR DESCRIPTION
**Pull request by ryanbogan**
Signed-off-by: Ryan Bogan <rbogan@amazon.com>

### Description
Replaces synchronous CountDownLatches with asynchronous CompletableFutures

### Issues Resolved
https://github.com/opensearch-project/opensearch-sdk-java/issues/292

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by saratvemulapalli — DISMISSED** [review]

**Review by dbwiddis — DISMISSED** [review]

**Review by dbwiddis — APPROVED** [review]
I before E, except after code review...

File: src/main/java/org/opensearch/sdk/handlers/ClusterStateResponseHandler.java [hunk 1]
```diff
@@ -68,11 +68,14 @@ public ClusterStateResponse read(StreamInput in) throws IOException {
...
      * Invokes await on the ClusterStateResponseHandler count down latch
-     * @throws InterruptedException
+     * @throws Exception
      *     if the response times out
      */
-    public void awaitResponse() throws InterruptedException {
-        inProgressLatch.await(ExtensionsManager.EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS);
+    public void awaitResponse() throws Exception {
+        inProgressFuture.orTimeout(ExtensionsManager.EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS);
+        if (inProgressFuture.isCompletedExceptionally()) {
```
Comment (owaiskazi19) [inline src/main/java/org/opensearch/sdk/handlers/ClusterStateResponseHandler.java:76]: Just for my understanding why are we checking this condition? Wouldn't `get` take care of it?

**Review by owaiskazi19 — DISMISSED** [review]

File: src/main/java/org/opensearch/sdk/handlers/ClusterStateResponseHandler.java (hunk 1 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/handlers/ClusterStateResponseHandler.java:76]: I think we actually are going to change this, pending CR of [#5646](https://github.com/opensearch-project/OpenSearch/pull/5646).   

TLDR: we need to use either `.get()` or `.join()` in all cases or it won't behave as we expect.

**Review by dbwiddis — CHANGES_REQUESTED** [review]
(Undoing approval as I'm aware we need to make changes corresponding to review of https://github.com/opensearch-project/OpenSearch/pull/5646)

File: src/main/java/org/opensearch/sdk/handlers/ClusterStateResponseHandler.java (hunk 1 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/sdk/handlers/ClusterStateResponseHandler.java:76]: Please change this to:
```java
inProgressFuture.orTimeout(ExtensionsManager.EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS).get();
```
and remove the next 3 lines.

File: src/main/java/org/opensearch/sdk/handlers/EnvironmentSettingsResponseHandler.java [hunk 2]
```diff
@@ -68,11 +68,14 @@ public EnvironmentSettingsResponse read(StreamInput in) throws IOException {
...
     /**
      * Invokes await on the EnvironmentSettingsResponseHandler count down latch
-     * @throws InterruptedException
+     * @throws Exception
      *        if the response times out
      */
-    public void awaitResponse() throws InterruptedException {
-        inProgressLatch.await(ExtensionsManager.EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS);
+    public void awaitResponse() throws Exception {
+        inProgressFuture.orTimeout(ExtensionsManager.EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/sdk/handlers/EnvironmentSettingsResponseHandler.java:75]: Please change this to:
```java
inProgressFuture.orTimeout(ExtensionsManager.EXTENSION_REQUEST_WAIT_TIMEOUT, TimeUnit.SECONDS).get();
```
and remove the next 3 lines.

**Review by dbwiddis — CHANGES_REQUESTED** [review]

**Review by dbwiddis — APPROVED** [review]

**Review by saratvemulapalli — APPROVED** [review]
